spec/keycard/institution_finder_spec.rb

```diff
@@ -0,0 +1,99 @@
+# frozen_string_literal: true
+
+require "keycard/institution_finder"
+require "sequel_helper"
+require "ipaddr"
+require "pry"
+
+RSpec.describe Keycard::InstitutionFinder do
+  subject { described_class.new }
+
+  describe "#attributes_for" do
+    let(:request) { double(:request) }
+
+    def add_inst_network(db, inst:, network:, access:)
+      @unique_id ||= 0
+      @unique_id += 1
+      range = IPAddr.new(network).to_range
+      db[:aa_network].insert([@unique_id,nil,network,range.first.to_i,range.last.to_i,access.to_s,nil,inst,Time.now,'test','f'])
+    end
+
+    before(:all) do
+      @db = test_database
+
+      add_inst_network(@db, inst: 1, network: '10.0.0.0/16', access: :allow)
+      add_inst_network(@db, inst: 1, network: '10.0.2.0/24', access: :deny)
+
+      # range in two institutions
+      add_inst_network(@db, inst: 2, network: '10.0.1.0/24', access: :allow)
+
+      # denied from one, allowed to another
+      add_inst_network(@db, inst: 1, network: '10.0.3.0/24', access: :deny)
+      add_inst_network(@db, inst: 2, network: '10.0.3.0/24', access: :allow)
+    end
+
+    before(:each) do
+      allow(request).to receive(:get_header)
+        .with('X-Forwarded-For')
+        .and_return(client_ip)
+    end
+
+    subject { described_class.new(@db).attributes_for(request) }
+
+    context "with an ip with a single institution" do
+      let(:client_ip) { "10.0.0.1" }
+
+      it "returns a hash with (only) a dlpsInstitutionIds key" do
+        expect(subject.keys).to contain_exactly('dlpsInstitutionIds')
```

Comment:
One thing I've been exploring is to avoid bare `subject`. I don't think the BDD aspects really suffer from having a domain-oriented name in the expectations. I don't mind this style -- just playing around with some variations to start building recommendations.